Reimplement usage of Google Translate API as a named utility to be able to provide different services

erral opened the pull request:
Addresses #467 

## Motivation

The current implementation requires anyone wanting to provide some other external translation service, to override the existing view and do the logic there.

I think that the implementation proposed here is flexible enough to be able to register several different services

## Introduced components

A new interface `IExternalTranslationService` is introduced which has several attributes and methods. This interface should be implemented by a utility registered using `zcml` for each translation service you want to have.

- `is_available()`: to return whether the adapter is available (for instance, is the Google translation API key entered into the control panel?)
- `available_languages()`: a list of language pairs (source, target), that this service supports.
- `order`: the order in which this adapter will be executed. This way, one can prioritize some services over others.
- `translate_content(content, source, target)`: the actual function that does the translation call.

I have converted the function that currently calls Google Translate into a utility providing `IExternalTranslationService`

## Strategy

The current browser view that does the call to Google Translate has been rewritten to get all registered utilities that are available.

Then it checks that the adapter supports the language-pair (source, target) translation.

If so, it requests the translation and returns it.

## Other solutions I thought of

- I thought also on creating a ZCML tag just for registering this kind of adapters, but found easier going the adapter way.
- Another way could be to register adapters instead of utilities.

## Documentation

If this is accepted, a documentation PR will be issued from this branch, where this adapter registration has been documented:

Branch: https://github.com/plone/documentation/tree/erral-adapter-registration
Documentation of the adapter: https://github.com/plone/documentation/blob/erral-adapter-registration/docs/i18n-l10n/use-an-external-translation-service.md#using-other-translation-services

## Other

I will update the PR with the changes in #462 if/when merged

erral committed:
convert the function that uses google translate to an adapter based lookup, so that extra services can be registered

erral committed:
implement adapter ordering

erral committed:
implement availability of service

erral committed:
changelog

mister-roboto commented:
@erral thanks for creating this Pull Request and helping to improve Plone!

TL;DR: Finish pushing changes, pass all other checks, then paste a comment:
```
@jenkins-plone-org please run jobs
```

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass, but it takes 30-60 min.  Other CI checks are usually much faster and the Plone Jenkins resources are limited, so when done pushing changes and all other checks pass either [start all Jenkins PR jobs yourself](https://jenkinsploneorg.readthedocs.io/en/latest/run-pull-request-jobs.html#run-a-pull-request-job), or simply add the comment above in this PR to start all the jobs automatically.

Happy hacking!

erral committed:
black

erral committed:
isort

erral committed:
format

erral committed:
remove empty

erral commented:
@jenkins-plone-org please run jobs

erral committed:
tests

erral committed:
isort

erral commented:
@jenkins-plone-org please run jobs

erral committed:
isort

erral committed:
lint

erral committed:
lint

erral commented:
@jenkins-plone-org please run jobs

erral committed:
Merge branch 'master' into erral-issue-467

tisto commented:
@erral  is this an enhancement or a PLIP?

erral commented:
As discussed, I don't know if this in itself is a PLIP... it is definetly an enhancement.

Something worth a PLIP would be to integrate these translations options in Volto.

I will add a REST API endpoint to this PR in order to be able to use these translation services, which would definetely help in the volto implementation.

erral committed:
convert to utilities

erral committed:
changelog

erral committed:
remove unneeded·

mamico commented on src/plone/app/multilingual/browser/translate.py:
```diff
@@ -1,53 +1,53 @@
+            not available_languages
+            or (source_language, target_language) in available_languages
+        ):
+            translation = adapter.translate_content(
```
IMHO it is better to handle this loop using a try/except pattern, so that the whole service is not repeated if you try to translate something like empty text.

mamico commented on src/plone/app/multilingual/google_translate.py:
```diff
@@ -0,0 +1,71 @@
+class GoogleCloudTranslationAPIFactory:
+    """implement the external translation using Google Cloud Translation API"""
+
+    order = 999
```
An idea might be to define a schema for translation service controlpanel with “enabled”, “order”, “available_source_languages”, and “available_target_languages” fields.

erral replied:
Yes, I tried to keep backwards compatibility in thilese changes and that's why I didn't change the controlpanel at all.

But what you say makes total sense. The point is: should we extract the google translate support from p.a.multilingual and put it in an addon?

jensens replied:
I think having the translation services as add-ons only totally makes sense. These days there are more and better services than Google available (opposed to the era the feature origins in).

mamico replied:
@erral @jensens @tkimnguyen I have a question about this. Currently, PAM has a Google Translate implementation, but I believe it is not working, anymore. Fix it and leaving Google as the only option, rather than considering other available implementations, seems unfair to me. Should we leave PAM without any translation implementation, or can we integrate something like LibreTranslate, which, unlike purely proprietary SaaS solutions, also provides an open-source implementation that can be hosted on our own server?

(By the way, when I find a free day, I'll wrap up everything from https://github.com/mamico/collective.translators and move it to Collective.)"

erral committed:
only translate non-empty values

erral committed:
Merge branch 'master' into erral-issue-467

erral committed:
add plone.restapi test requirements

erral commented:
@jenkins-plone-org please run jobs

erral committed:
Merge branch 'master' into erral-issue-467

wesleybl committed:
Adds the volto.blocks behavior to LRF if `plone.volto` is installed.

During the installation of `plone.app.multilingual`, checks if `plone.volto` is installed.
If it is, adds the volto.blocks behavior to LRF.

wesleybl committed:
Move the function `add volto blocks_behavior_to_lrf` to the class `Setup Multilingual Site`.

erral committed:
Merge branch 'master' into erral-issue-467

erral committed:
zpretty

erral committed:
fix dependnecies

erral committed:
breaking

erral committed:
utlity

erral committed:
lint

erral committed:
mark changes as breaking

erral committed:
remove more google traces

erral committed:
lint

erral commented:
@jenkins-plone-org please run jobs

erral committed:
fixes

erral committed:
remove unneeded

erral committed:
remove unneeded

erral committed:
remove unneeded

erral committed:
test new endpoint

erral committed:
rename

erral committed:
lint

erral committed:
add missing test-dependency

erral commented:
@jenkins-plone-org please run jobs

erral commented:
Following my latest comment, I have reworked this PR a bit, completely removing the Google Translate functionality from here and renaming the view that is called from the `babel_view`. 

Because of that breaking change, I have added a PLIP explaining the motivation and the deliverables:

https://github.com/plone/Products.CMFPlone/issues/4244

erral committed:
load plone.rest to have the plone:service registration available

erral committed:
register the endpoints only when p.a.multilingual is installed

erral committed:
setup.py

erral committed:
remove unneeded

erral committed:
load plone.restapi for tests

erral committed:
fix: load controlpanel permissions

erral committed:
remove gtranslate icon and use plone provided translate icon

erral committed:
restore tinymce detection

erral committed:
Merge branch 'master' into erral-issue-467

erral committed:
zpretty

mamico committed:
Merge branch 'master' into erral-issue-467

mamico commented:
@jenkins-plone-org please run jobs

mamico commented:
@erral Just checking in on this PR since it's been open for a while. I'm about to deploy these changes to production for a client (classic ui, not volto), so I wanted to see if everything is good on your end and if you're planning to merge it soon.
There's no rush at all on my side since I can easily work with the code from this branch. Let me know if you need anything from me for the review! Thanks

erral commented:
I didn't reach on time for 6.2, so I think I will implement a backwards-compatibility layer on this PR, to be accepted.

Anyway, can you second the PLIP (https://github.com/plone/Products.CMFPlone/issues/4244) and this way perhaps we can expedite merging it. Thx!

erral committed:
Restore backward compatibility for Google Translate service

erral committed:
do not force to use the google translation service

erral committed:
rename changelog

erral committed:
changelog

erral committed:
zpretty

erral committed:
Merge branch 'master' into erral-issue-467

erral commented:
@jenkins-plone-org please run jobs

erral commented:
@mamico I have updated the PR keeping backwards compatibility (I mean, keeping the Google translate configuration et al, but I have named the utility *differently* from the one in collective.translators to avoid any clashes.)

Tested it locally and it works nicely.

We will have a client from September on where we will have to connect Plone to an internal translation API, so having this will be really handy!

erral committed:
provide testing utilities

erral committed:
missing imports

erral commented:
@jenkins-plone-org please run jobs
